sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -644,40 +645,97 @@ abstract class RapidsShuffleThreadedReaderBase[K, C](
+              // this is stored only to call close on it
+              currentStream = stream
+              currentIter = serializerInstance.deserializeStream(stream).asKeyValueIterator
+            }
+            val res = currentIter.next()
+            val fetchTime = System.nanoTime() - fetchTimeStart
+            deserializationTimeNs.foreach(_ += (fetchTime - readBlockedTime))
+            shuffleReadTimeNs.foreach(_ += fetchTime)
+            res
           }
-          val res = currentIter.next()
-          val fetchTime = System.nanoTime() - fetchTimeStart
-          deserializationTimeNs.foreach(_ += (fetchTime - readBlockedTime))
-          shuffleReadTimeNs.foreach(_ += fetchTime)
-          res
         }
+      } else {
+        null
       }
-    } else {
-      null
-    }
 
-    // Register a completion handler to close any queued cbs.
+    // Register a completion handler to close any queued cbs,
+    // pending iterators, or futures
     onTaskCompletion(context) {
+      // remove any materialized batches
       queued.forEach {
         case (_, cb:ColumnarBatch) => cb.close()
       }
       queued.clear()
+
+      // close any materialized BlockState objects that are holding onto netty buffers or
+      // file descriptors
+      pendingIts.safeClose()
+      pendingIts.clear()
+
+      // we could have futures left that are either done or in flight
+      // we need to cancel them and then close out any `BlockState`
+      // objects that were created (to remove netty buffers or file descriptors)
+      val futuresAndCancellations = futures.map { f =>
+        val didCancel = f.cancel(true)
+        (f, didCancel)
+      }
+
+      // if we weren't able to cancel, we are going to make a best attempt at getting the future
+      // and we are going to close it. The timeout is to prevent an (unlikely) infinite wait.
+      // If we do timeout then this handler is going to throw.
+      var failedFuture: Option[Throwable] = None
+      futuresAndCancellations
+        .filter { case (_, didCancel) => !didCancel }
+        .foreach { case (future, _) =>
+          try {
+            // this could either be a successful future, or it finished with exception
+            // the case when it will fail with exception is when the underlying stream is closed
+            // as part of the shutdown process of the task.
+            future.get(10, TimeUnit.MILLISECONDS)
+              .foreach(_.close())
+          } catch {
+            case t: Throwable =>
+              // this is going to capture the first exception and not worry about others
+              // because we probably don't want to spam the UI or log with an exception per
+              // block we are fetching
+              if (failedFuture.isEmpty) {
+                failedFuture = Some(t)
+              }
+          }
+        }
+      futures.clear()
+      if (fallbackIter != null) {
+        fallbackIter.close()
```

Comment:
if this throws and failedFuture was not empty at the time, we are potentially masking the root cause

Reply:
ah good point, I can add it as suppressed, sec.

Reply:
@gerashegalov should be taken care of here: https://github.com/NVIDIA/spark-rapids/pull/10637/commits/d58b4711db8fb59b5082d4172df91a27312c3519